### ISSUE TITLE
Segmentation fault using page.getLinks() on encrypted PDF

### ISSUE THREAD

**Issue by timb07** [rk700/PyMuPDF#179]
I'm extracting URIs from link annotations in PDFs. While I don't expect to get anything useful from an "encrypted" PDF without supplying a password, I don't expect a segmentation fault either:

```py
Python 3.5.2 (default, Sep 10 2016, 08:21:44) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fitz, faulthandler
>>> faulthandler.enable()
>>> doc = fitz.open('test_encrypted.pdf')
>>> for page in doc: page.getLinks()
... 
Fatal Python error: Segmentation fault

Current thread 0x00007f3a83764700 (most recent call first):
  File "/home/tbell/.virtualenvs/xxx/lib/python3.5/site-packages/fitz/fitz.py", line 749 in resolveLink
  File "/home/tbell/.virtualenvs/xxx/lib/python3.5/site-packages/fitz/fitz.py", line 2854 in dest
  File "/home/tbell/.virtualenvs/xxx/lib/python3.5/site-packages/fitz/utils.py", line 176 in getLinkDict
  File "/home/tbell/.virtualenvs/xxx/lib/python3.5/site-packages/fitz/utils.py", line 233 in getLinks
  File "<stdin>", line 1 in <module>
Segmentation fault
```

This is with PyMuPDF==1.13.7.

I'm not sure whether it's an issue with the underlying MuPDF library, or whether PyMuPDF should just not try to use `resolveLink()` on an encrypted PDF.

[test_encrypted.pdf](https://github.com/rk700/PyMuPDF/files/2109555/test_encrypted.pdf)

**Comment by JorjMcKie** [comment]
It is an MuPDF error. While it is true that I forgot to insert an "fz_try - fz_catch" sequence here, it didn't prevent the segfault after I corrected this.

`page.getLinks()` does work (in your case at least) if I return `None` from `resolveLink`:
```python
>>> page.getLinks()
[{'kind': 3, 'xref': 3, 'from': fitz.Rect(78.0, 77.25, 97.44999694824219, 89.25),
'file': 'H\udc92\udc85P(\udcd8\x1eRP+@\udca4\udcc7\udcff\udcd68I\udc96\udcf3a\udca8\udc85\udcc2'}]
```
But (except for the `fitz.Rect`) the result is meaningless anyway, because MuPDF determines the type by parsing the `uri` string - which is encrypted stuff of course.

Similar nonsense happens at other places like pixmap creation ...
So what I will do is treating non-decrypted documents as if they where closed.
Any comments?

**Comment by timb07** [comment]
> So what I will do is treating non-decrypted documents as if they where closed.

That sounds fine. In the meantime, I have added a workaround in my own code to check if the document is encrypted, and not use `getLinks()` if so.

However, I have just encountered a PDF where my workaround does not prevent a segmentation fault; the Python traceback is exactly the same as above. But I can open this PDF without entering a password, although it does have an Encryption object. (I don't understand all the intricacies of PDF encryption and permissions/access control.) So I suspect your proposed solution won't address this particular case; it may require a fix in MuPDF.

[x1.pdf](https://github.com/rk700/PyMuPDF/files/2113314/x1.pdf)

**Comment by timb07** [comment]
Re the junk characters at the end of the URI string, isn't it possible that they're present in the encrypted string? I counted the characters in the encrypted URI strings, and the lengths are the same as the decrypted URIs, complete with trailing junk characters. So it could be that this PDF is bad itself, rather than how MuPDF is handling it.

And yes, please do include the example file in the bug report to MuPDF. Thanks. :)

**Comment by JorjMcKie** [comment]
Ok, will do then.

Counter-arguments:
- Viewer software like Adobe Reader displays the page correctly, including the 3 links. 
- looking at this PDF's "source" with a suitable editor shows that the URI strings (object xrefs 18, 19, and 29) are encrypted completely (contain no directly readable part). So, because I get something partly readable, decryption has at least somehow worked.

Anyway, let's see how Artifex reacts ...

**Comment by JorjMcKie** [comment]
I have just uploaded v1.13.11.
You can find it in the releases dir of this repo and also on PyPI within the next 5 minutes.
This version should fix your issue as discussed above.
If you can confirm this, please close the issue or leave me a hint - thanks!

**Comment by timb07** [comment]
Thanks for your quick response and all your work on this issue. Yes, the problem seems to be fixed in 1.13.11.
